### ISSUE TITLE
Incremental compilation

### ISSUE THREAD

**Issue by ajoberstar** [gradle-clojure/gradle-clojure#16]
As of 0.1.0, gradle-clojure partially participates in the incremental build feature of Gradle. This is where inputs and outputs of a task are declared, and if either changes the whole task will re-run.

We have not yet annotated all of our task properties to participate in this fully.

In addition, to have true incremental compilation we need to use a task that takes an [IncrementalTaskInputs](https://docs.gradle.org/current/userguide/custom_tasks.html#incremental_tasks), which lists each file that changed or was removed from the inputs.

I think by combining that information with [tools.namespace](https://github.com/clojure/tools.namespace) for detecting namespaces in files and understanding the dependency tree between the namespaces, we should be able to do this incrementally. The main challenge is cleaning the classes that were output.

**Comment by pbzdyl** [comment]
I think we could start with the simple version of incremental compilation that would help in multi-module Clojure projects (that was implemented in [my PR to cursive-ide/gradle-clojure](https://github.com/cursive-ide/gradle-clojure/pull/12)):
- if there was no change in the Clojure compile task inputs, the task won't be execute by Gradle at all
- if there was a change in the inputs detected by Gradle, perform a full compilation

That would make sure that only modules with changes in the compile task inputs would be fully recompiled.

For this to happen we need the following changes in the compile task:
1. properly mark all the inputs
2. provide task method which accepts `IncrementalTaskInputs` argument (but it would be unused as the task would always execute a full compilation)
3. implement a way to delete class files generated by the compile task

[I have implemented #3](https://github.com/pbzdyl/gradle-clojure/blob/issues/10/src/main/kotlin/cursive/ClojurePlugin.kt#L296-L308) in [my PR to cursive-ide/gradle-clojure](https://github.com/cursive-ide/gradle-clojure/pull/12) by using a temporary directory where compiled classes written and once a compilation is done they are copied to the source set's destination directory. During the next compilation (if it is an incremental compilation as indicated by `IncrementalTaskInputs` argument data), all the files found in the temporary directory would be deleted from source set's destination directory before running another full compilation.

Would this approach be OK to start with? If yes, I could help by submitting a PR implementing it.

**Comment by ajoberstar** [comment]
Yeah, I agree this is the behavior to start with. It's partially in place already (I ported your code and tests as part of 0.1.0), see my notes below for what's still missing.

> 1. Properly mark all the inputs

The source files get auto marked as inputs due to extending `AbstractCompile`. I didn't get through marking the rest of the inputs, though. If you run `./gradlew validateTaskProperties` it will list the remaining ones.

> 2. provide task method which accepts IncrementalTaskInputs argument (but it would be unused as the task would always execute a full compilation)

The up-to-date style incremental behavior doesn't use this method, from my understanding. At least for the source files changing, I've verified that it correctly shows UP-TO-DATE and skips the task. We wouldn't need to add this method until we try to go full incremental.

> 3. Implement a way to delete class files generated by the compile task

I originally had kept all of your behavior for this, but removed it in the changes planned for 0.2.0. This was because we now generate Clojure class output into a different directory than Java output, so it doesn't need to be further isolated. We probably need to add back deleting the output directory at the beginning of the task, I didn't verify whether that was working correctly.

It would be great if you could put a PR together for this!
